tensorboard/loader.py

```diff
@@ -399,6 +401,254 @@ def __str__(self):
     return u'BufferedRecordReader{%s}' % self.path
 
 
+class RateCounter(object):
+  """Utility class for tracking how much a number increases each second.
+
+  The rate is calculated by averaging of samples within a time window,
+  which weights recent samples more strongly.
+  """
+
+  def __init__(self, window, clock=time.time):
+    """Creates new instance.
+
+    Args:
+      window: The maximum number of seconds across which rate is
+          averaged. In practice, the rate might be averaged over a time
+          period greater than window if set_value is being called less
+          frequently than window.
+      clock: Function returning a float with the number of seconds
+          since the UNIX epoch in zulu time.
+
+    :type window: float
+    :type clock: () -> float
+    """
+    self._window = window
+    self._clock = clock
+    self._points = collections.deque()
+    self._last_value = None  # type: int
+    self._last_time = None  # type: float
+
+  def get_rate(self):
+    """Determines rate of increase in value per second averaged over window.
+
+    Returns:
+      An integer representing the rate or None if not enough
+      information has been collected yet.
+
+    :rtype: int
+    """
+    toto = 0.0
+    points = []
+    for i, (rate, weight, _) in enumerate(self._points):
+      weight *= 1.0 / (i + 1)
```

Comment:
Just a note: this is an index-weight, not a time-weight. (That is, the semantics are different from using `weight *= 1.0 / (time - self._last_time)`.) It's not clear to me what advantages either has over the other.